### PR TITLE
New version notifier

### LINKED ISSUE #1312: Customer gets notified about new Kyma CLI version
**Description**

The Kyma CLI has to inform customers about new CLI releases. He gets informed by showing a notification message on the console output (message is shown before any command output is printed).

AC:
* When the CLi is called, it checks for a new version and notifies the customer about it (info-message at the beginning of the console output)
* Customer can suppress the version-check by setting an additional flag (e.g.  `--no-version-check`)

**Reasons**
Hold customers up-to-date and to ensure they are using latest patched Kyma CLI version.

**Attachments**

### PR DESCRIPTION
**Pull request by amritanshusikdar**
**Description**
User gets notified if using an old version of the CLI with a warning to upgrade to a newer version.

**Related issue(s)**

Resolves #1312

### CODE REVIEW

File: cmd/kyma/version/checker.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+		for _, setting := range info.Settings {
+			if setting.Key == "vcs.revision" {
+				currentCommitSHA = setting.Value
+			}
+		}
+	}
+
+	nicePrint := nice.Nice{}
+	if Version != stableCLINumber || currentCommitSHA != stableCLI.Commit.SHA {
+		nicePrint.PrintImportantf("CAUTION: You're using an outdated version of the Kyma CLI. The latest stable version is: %s", stableCLINumber)
```
Comment (lindnerby) [inline cmd/kyma/version/checker.go:66]: Please print the current version as well

File: cmd/kyma/version/checker.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+	var currentCommitSHA string
+	if info, ok := debug.ReadBuildInfo(); ok {
+		for _, setting := range info.Settings {
+			if setting.Key == "vcs.revision" {
+				currentCommitSHA = setting.Value
+			}
+		}
+	}
+
```
Comment (lindnerby) [inline cmd/kyma/version/checker.go:63]: Let's return if any value required for the condition check is empty, although it's very unlikely to happen

File: cmd/kyma/version/checker.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	var stableCLINumber string
+	for _, tag := range githubTags {
+		if tag.Name != stableCLI.Name && tag.Commit.SHA == stableCLI.Commit.SHA {
+			stableCLINumber = tag.Name
+		}
+	}
+
+	var currentCommitSHA string
+	if info, ok := debug.ReadBuildInfo(); ok {
```
Comment (lindnerby) [inline cmd/kyma/version/checker.go:56]: Let's early return if `!ok`

**Review by lindnerby — CHANGES_REQUESTED** [review]

File: cmd/kyma/version/checker.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/kyma-project/cli/internal/nice"
+	"io"
+	"net/http"
+	"runtime/debug"
+)
+
+type KymaCLIMetadata struct {
```
[the remark this answers is not in the document]
Comment (amritanshusikdar) [reply cmd/kyma/version/checker.go:12]: Going ahead with `LatestGitTag`

File: cmd/kyma/version/checker.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+		SHA string `json:"sha"`
+	} `json:"commit"`
+}
+
+// CheckForStableRelease Checks for new versions of the CLI
+func CheckForStableRelease() {
+	response, err := http.Get("https://api.github.com/repos/kyma-project/cli/tags")
+
+	//	For any problems in fetching, reading or parsing the response from GitHub API, we simply ignore it
+	//	and don't disrupt the usual CLI Flow
```
[the remark this answers is not in the document]
Comment (amritanshusikdar) [reply cmd/kyma/version/checker.go:24]: done

File: cmd/kyma/version/checker.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+	"encoding/json"
+	"fmt"
+	"github.com/kyma-project/cli/internal/nice"
+	"io"
+	"net/http"
+	"regexp"
+)
+
+const (
+	GitHubAPIEndpoint = "https://api.github.com/repos/kyma-project/cli/releases/latest"
```
Comment (lindnerby) [inline cmd/kyma/version/checker.go:13]: Can be pkg-private

File: cmd/kyma/version/checker.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+)
+
+type LatestGitTag struct {
```
Comment (lindnerby) [inline cmd/kyma/version/checker.go:16]: Can be pkg-private

File: cmd/kyma/version/checker.go [hunk 8]
```diff
@@ -0,0 +1,46 @@
+package version
```
Comment (lindnerby) [inline cmd/kyma/version/checker.go:1]: Probably not the fitting pkg for this function. Can we move it to somewhere internal?

File: cmd/kyma/version/checker.go [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+		return
+	}
+
+	var latestGitTag LatestGitTag
+	if err := json.Unmarshal(responseData, &latestGitTag); err != nil {
+		return
+	}
+
+	matched, err := regexp.MatchString("[0-9]+[.][0-9]+[.][0-9]+", Version)
+	if err != nil {
```
Comment (lindnerby) [inline cmd/kyma/version/checker.go:38]:
```suggestion
	if err != nil || !matched {
```

Comment (lindnerby) [reply cmd/kyma/version/checker.go:38]: Let's semantically group the check here

**Review by lindnerby — CHANGES_REQUESTED** [review]

**Review by lindnerby — APPROVED** [review]

**Review by mmitoraj — APPROVED** [review]

**Pull request merged**